File: boosting/src/data/dataholder/DoubleDataHolder.java
package data.dataholder;

import data.DataReader;

public class DoubleDataHolder extends DataHolder<Double> {

    public DoubleDataHolder(DataReader<Double> dataReader, double trainPercent, boolean normalize, boolean shuffle) {
        super(dataReader, trainPercent, normalize, shuffle);
    }

    @Override
    protected void normalize() {
        if (data == null || data.isEmpty()) {
            return;
        }
        double min = Double.POSITIVE_INFINITY, max = Double.NEGATIVE_INFINITY;
        for (double x : data) {
            min = Math.min(x, min);
            max = Math.max(x, max);
        }
        for (int i = 0; i < data.size(); i++) {
            data.set(i, (data.get(i) - min) / (max - min));
        }
    }

    @Override
    public int getVectorSize() {
        return 1;
    }
}

File: boosting/src/data/dataholder/DoubleDataHolder.kt
package data.dataholder

import data.DataReader

class DoubleDataHolder(dataReader: DataReader<Double>, trainPercent: Double, normalize: Boolean, shuffle: Boolean) : DataHolder<Double>(dataReader, trainPercent, normalize, shuffle) {

    override fun normalize() {
        var min = Double.POSITIVE_INFINITY
        var max = Double.NEGATIVE_INFINITY
        for (x in data) {
            min = Math.min(x, min)
            max = Math.max(x, max)
        }
        data.forEach { x -> x - min }
        if (min != max) {
            data.forEach { x -> x / (max - min) }
        }
    }

    override val vectorSize: Int
        get() = 1
}
